nengo_spinnaker/netlist/utils.py

```diff
@@ -0,0 +1,169 @@
+    subvertex. Later this ID can be used in the keyspace of all nets
+    originating from the cluster.
+
+    Parameters
+    ----------
+    groups : [{Vertex, ...}, ...]
+        List of groups of vertices. This is used to identify objects which may
+        be joined into a cluster.
+    placements : {vertex: (x, y), ...}
+    """
+    # Perform clustering for each group of vertices in turn.
+    for group in groups:
+        # Build a map of placements to vertices, this will be used to identify
+        # clusters.
+        clusters = collections.defaultdict(list)  # Map co-ordinate to cluster
+
+        # Add each vertex in the group to a cluster.
+        for vertex in group:
+            coord = placements[vertex]
+            clusters[coord].append(vertex)
+
+        # Assign a unique ID to each cluster in the group
+        for cluster_id, cluster in enumerate(itervalues(clusters)):
+            # Store the cluster ID in each vertex in the cluster
+            for vertex in cluster:
+                vertex.cluster = cluster_id
+
+
+def get_net_keyspaces(placements, derived_nets):
+    """Get a map from the nets used during routing to the keyspaces (NOT the
+    keys and masks) that should be used when building routing tables.
+
+    Cluster IDs will be applied to any nets which used the default Nengo
+    keyspace.
+
+    Parameters
+    ----------
+    placements : {vertex: (x, y), ...}
+    derived_nets : {:py:class:`~nengo_spinnaker.netlist.Net`:
+                    {(x, y): :py:class:`~rig.netlist.Net`, ...}, ...}
+        Map from original nets to co-ordinates and the derived nets which
+        originate from them as, returned by :py:func:`~.get_routing_nets`.
+
+    Returns
+    -------
+    {net: keyspace, ...}
+        A map from nets to :py:class:`~rig.bitfield.BitField`s that can later
+        be used to generate routing tables.
+    """
+    net_keyspaces = dict()  # Map from derived nets to keyspaces
+
+    for original_net, placement_nets in iteritems(derived_nets):
+        for placement, net in iteritems(placement_nets):
+            # If the keyspace is the default Nengo keyspace then add a cluster
+            # ID, otherwise just store the keyspace as is.
+            if is_nengo_keyspace(original_net.keyspace):
+                # Get all the cluster IDs assigned to vertices with the given
+                # placement (there should only be one cluster ID, if there are
+                # more it would imply that multiple Nengo objects ended up in
+                # the sources for a given Net and it is an error from which we
+                # cannot recover).
+                cluster_ids = set(vx.cluster for vx in original_net.sources
+                                  if placements[vx] == placement)
+                assert len(cluster_ids) == 1, "Inconsistent cluster IDs"
+                cluster_id = next(iter(cluster_ids)) or 0  # Get the ID
```

Comment:
It is possible for `cluster_ids` to equal `{None}`; the 0 catches this case.  If `cluster_ids` were empty then `next(iter(cluster_ids))` would raise `StopIteration` and the `or` wouldn't save us.